## Problem 9: Special Pythagorean Triplet

A Pythagorean triplet is a set of three natural numbers, $a < b < c$, for which,

$a^2 + b^2 = c^2$

For example, $3^2 + 4^2 = 9 + 16 = 25 = 5^2$.

There exists exactly one Pythagorean triplet for which $a + b + c = 1000$.
Find the product $abc$.

In [1]:
''' 
Checks if numbers form a pythagorean triplet.

eg:
>>> is_pythagorean_triplet(3, 4, 5)
>>> True
'''

def is_pythagorean_triplet(a : int, b : int, c : int) -> bool:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Checks if numbers form a pythagorean triplet.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : a : The 1st number in the triplet.
        : int : b : The 2nd number in the triplet.
        : int : c : The 3rd number in the triplet.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : bool : The numbers form a Pythagorean Triplet.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> is_pythagorean_triplet(3, 4, 4)
        >>> True
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(a, int): raise ValueError('Please enter an integer > 0 for the a argument.')
    if a < 1: raise ValueError('Please enter an integer > 0 for the a argument.')
    if not isinstance(b, int): raise ValueError('Please enter an integer > 0 for the b argument.')
    if b < 1: raise ValueError('Please enter an integer > 0 for the b argument.')
    if not isinstance(c, int): raise ValueError('Please enter an integer > 0 for the c argument.')
    if c < 1: raise ValueError('Please enter an integer > 0 for the c argument.')
    
    return (a**2 + b**2 == c**2)

In [2]:
''' 
Searches through all possible combinations to find all Pythagorean Triplet.
0 < a < target_sum / 3 since a < b < c
a < b < target_sum / 2 since b < c
c = target_sum - a - b since a + b + c = target_sum

eg:
>>> search_numbers(1000)
>>> [[200, 375, 425]]
'''

def search_numbers(target_sum : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Searches through all possible combinations to find all Pythagorean Triplet.
        0 < a < target_sum / 3 since a < b < c
        a < b < target_sum / 2 since b < c
        c = target_sum - a - b since a + b + c = target_sum
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : target_sum : The target sum of the triplet.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : The numbers form a Pythagorean Triplet and sum to the target.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> search_numbers(1000)
        >>> T[[200, 375, 425]]rue
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(target_sum, int): raise ValueError('Please enter an integer > 0 for the target_suma argument.')
    if target_sum < 1: raise ValueError('Please enter an integer > 0 for the target_sum argument.')
    
    # Store all solutions in list
    solutions = []
    
    # 0 < a < target_sum / 3 since a < b < c
    for a in range(1, int((target_sum)/3)+1):
        
        # a < b < target_sum / 2 since b < c
        for b in range(a+1, int((target_sum+1)/2)+1):
            
            # c = target_sum - a - b since a + b + c = target_sum
            c = target_sum - a - b
            
            # Check if the set is a pythagorean triple.
            if is_pythagorean_triplet(a, b, c): solutions.append([a, b, c])
    
    # Return all solutions.
    return solutions

In [3]:
target_sum = 1000
problem_solution = search_numbers(target_sum)[0]

In [4]:
print(f'The Pythagorean triple that sums to {target_sum:,}:')
print(',    '.join(str(number) for number in problem_solution))
print(f'{problem_solution[0]}   + {problem_solution[1]}   + {problem_solution[2]}   = {problem_solution[0] + problem_solution[1] + problem_solution[2]:,}')
print(f'{problem_solution[0]}   * {problem_solution[1]}   * {problem_solution[2]}   = {problem_solution[0] * problem_solution[1] * problem_solution[2]:,}')
print(f'{problem_solution[0]}^2 + {problem_solution[1]}^2 = {problem_solution[0]**2:,} + {problem_solution[1]**2:,} = {problem_solution[2]**2:,} =  {problem_solution[2]}^2')

The Pythagorean triple that sums to 1,000:
200,    375,    425
200   + 375   + 425   = 1,000
200   * 375   * 425   = 31,875,000
200^2 + 375^2 = 40,000 + 140,625 = 180,625 =  425^2


### Problem 9 Solution:

The Pythagorean triple that sums to 1,000:

$200,    375,    425$

$200   + 375   + 425   = 1,000$

$200   * 375   * 425   = 31,875,000$

$200^2 + 375^2 = 40,000 + 140,625 = 180,625 =  425^2$.